client/finality-grandpa/src/communication/gossip.rs

```diff
@@ -1320,6 +1372,16 @@ impl<Block: BlockT> GossipValidator<Block> {
 			}
 		};
 
+		// Prometheus metric recording.
+		if let (Some(metrics), Some(message_name)) = (&self.metrics, message_name) {
+			let action_name = match action {
+				Action::Keep(_, _) => "keep",
+				Action::ProcessAndDiscard(_, _) => "process_and_discard",
+				Action::Discard(_) => "discard",
+			};
+			metrics.messages_validated.with_label_values(&[message_name, action_name]).inc();
```

Comment:
I hope that this does not introduces performance problems?

Reply:
Valid concern. I am curious myself, so let's dig into this a bit.

Within the benchmark suit of the `prometheus` crate (we are using version `v0.7.0`) there is the [`bench_counter_with_label_values`](https://github.com/tikv/rust-prometheus/blob/v0.7.0/benches/counter.rs#L24) test which mirrors our use case here the best. This test increases a `CounterVec` with 3 labels in a tight loop. Each increase is one iteration.

On my CPU (`i7-8550U`) an iteration takes ~ `100ns` which one could interpret as: **Increasing a Prometheus counter with 3 labels takes around `100ns`**. (Given that the benchmark above runs in a tight loop on a single thread, one could argue that it does not mirror the amount of CPU cache invalidation and cache coherency traffic that we would see in  a real-world scenario.)

The `GosssipValidator::do_validate` method is called on each incoming gossip message. One could assume that each gossip message translates to an IP packet that needs to be received. Receiving (not the latency on the wire, but only the time within the machine) an IP packet would roughly be around `0.025 ms`. Ignoring all additional steps involved, which would probably each exceed the `0.025 ms`, e.g. transport layer decryption, SCALE decoding, ... one could say that: **The lower bound of the `do_validate` function is at least `0.025 ms`**.

Comparing the two assumptions: **Increasing the counter would make up `0.4 %` (`100 ns / 0.025 ms`) of the overall `do_validate` execution time.** I would say this is an overhead worth ignoring / paying.

Does this sound reasonable to you @bkchr?

Reply:
You had me with `100ns` :D